source/grid/manifold.cc

```diff
@@ -81,32 +99,19 @@ get_new_point (const std::vector<Point<spacedim> > &surrounding_points,
   Assert(std::abs(std::accumulate(weights.begin(), weights.end(), 0.0)-1.0) < tol,
          ExcMessage("The weights for the individual points should sum to 1!"));
 
-  // The number of points is small (at most 26 in 3D), therefore try to minize
-  // overhead and use a very simple search of O(N^2) to find the order in which
-  // to loop over points.
+  // First sort points in the order of their weights. This is done to
+  // produce unique points even if get_intermediate_points is not
+  // associative (as for the SphericalManifold).
   std::vector<unsigned int> permutation(n_points);
-  std::vector<bool> found(n_points,false);
+  for (unsigned int i=0; i<n_points; ++i)
+    permutation[i] = i;
 
-  unsigned int min_index = numbers::invalid_unsigned_int;
-  for (unsigned int i=0; i < n_points; ++i)
-    {
-      double min_weight = std::numeric_limits<double>::max();
+  const CompareWeights compare_weights(weights);
+  std::sort(permutation.begin(),
+            permutation.end(),
+            compare_weights);
```

Comment:
No need to create the `compare_weights` object and let it live to the end of the function. Rather, just create an anonymous object of this type in place of the last argument.